src/librustc_mir/interpret/const_eval.rs

```diff
@@ -20,77 +20,76 @@ use rustc_const_math::ConstInt;
-            )?;
-            tcx.interpret_interner.borrow_mut().cache(
-                cid,
-                PtrAndAlign {
-                    ptr: ptr.into(),
-                    aligned: !layout.is_packed(),
-                },
-            );
-            let cleanup = StackPopCleanup::MarkStatic(Mutability::Immutable);
-            let name = ty::tls::with(|tcx| tcx.item_path_str(instance.def_id()));
-            trace!("const_eval: pushing stack frame for global: {}", name);
-            ecx.push_stack_frame(
-                instance,
-                mir.span,
-                mir,
-                Place::from_ptr(ptr),
-                cleanup.clone(),
-            )?;
-
-            while ecx.step()? {}
-
-            // reinsert the stack frame so any future queries have the correct substs
-            ecx.push_stack_frame(
-                instance,
-                mir.span,
-                mir,
-                Place::from_ptr(ptr),
-                cleanup,
-            )?;
-        }
-        let value = tcx.interpret_interner.borrow().get_cached(cid).expect("global not cached");
-        Ok((value, instance_ty))
-    })();
-    (try, ecx)
+    if ecx.tcx.has_attr(instance.def_id(), "linkage") {
+        return Err(ConstEvalError::NotConst("extern global".to_string()).into());
+    }
+    // FIXME(eddyb) use `Instance::ty` when it becomes available.
+    let instance_ty =
+        ecx.monomorphize(instance.def.def_ty(tcx), instance.substs);
+    if tcx.interpret_interner.borrow().get_cached(cid).is_none() {
+        let mir = ecx.load_mir(instance.def)?;
+        let layout = ecx.layout_of(instance_ty)?;
+        assert!(!layout.is_unsized());
+        let ptr = ecx.memory.allocate(
+            layout.size.bytes(),
+            layout.align.abi(),
+            None,
+        )?;
+        tcx.interpret_interner.borrow_mut().cache(
+            cid,
+            PtrAndAlign {
+                ptr: ptr.into(),
+                aligned: !layout.is_packed(),
+            },
+        );
+        let cleanup = StackPopCleanup::MarkStatic(Mutability::Immutable);
+        let name = ty::tls::with(|tcx| tcx.item_path_str(instance.def_id()));
+        trace!("const_eval: pushing stack frame for global: {}", name);
+        ecx.push_stack_frame(
+            instance,
+            mir.span,
+            mir,
+            Place::from_ptr(ptr),
+            cleanup.clone(),
+        )?;
+
+        while ecx.step()? {}
+    }
+    let value = tcx.interpret_interner.borrow().get_cached(cid).expect("global not cached");
+    Ok((value, instance_ty))
 }
 
 pub fn eval_body_as_integer<'a, 'tcx>(
     tcx: TyCtxt<'a, 'tcx, 'tcx>,
     param_env: ty::ParamEnv<'tcx>,
     instance: Instance<'tcx>,
 ) -> EvalResult<'tcx, ConstInt> {
-    let (ptr_ty, ecx) = eval_body(tcx, instance, param_env);
+    let ptr_ty = eval_body(tcx, instance, param_env);
     let (ptr, ty) = ptr_ty?;
+    let ecx = mk_eval_cx(tcx, instance.substs, param_env);
```

Comment:
Is the new `substs` field on `EvalContext` just for this?

Reply:
jup. We used to push a stackframe just to have substs.

Reply:
I removed the substs part and just simplified the interface